### PR TITLE
react-hooks: Add useMessages hook

### PR DESCRIPTION
**Pull request by splindsay-92**
### Context

* [CHA-384]
* As part of the react hooks development, a new hook to interact with the `Messages` instance of a room was required.

### Description

* Added `useMessages` hook to provide access to the `Messages` instance. This allows the user to interact with the underlying `Messages` instance, send messages, query historical messages and subscribe to new messages.
  * An optional parameter can be supplied alongside a listener, this will allow messages up-until the point the listener was subscribed.
* Updated the demo app to use the new hook for messages handling and removed the old custom hook.
* Updated the readme to document hook usage.
* Added unit tests for the new hook.
* Added integration tests for the new hook.

### Checklist

* [x] QA'd by the author.
* [ ] Unit tests created (if applicable).
* [x] Integration tests created (if applicable).
* [x] Follow coding style guidelines found [here](https://github.com/ably/engineering/tree/main/best-practices).
* [x] TypeDoc updated (if applicable).
* [x] (Optional) Update documentation for new features.
* [x] Browser tests created (if applicable).
* [x] In repo demo app updated (if applicable).

### Testing Instructions (Optional)

1. Spin up a local instance of the demo app.
2. Ensure you can send messages.
3. In another window, ensure you receive the messages.
4. Refresh a window and check the historical messages have been loaded.

[CHA-384]: https://ably.atlassian.net/browse/CHA-384?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: demo/src/containers/Chat/Chat.tsx [hunk 1]
```diff
@@ -63,15 +69,16 @@ export const Chat = () => {
     window.location.reload();
   }
 
-  const scrollToBottom = () => {
-    messagesEndRef.current?.scrollIntoView({ behavior: 'smooth' });
-  };
+  // TODO Cleanup commented code
```
Comment (AndyTWF) [inline demo/src/containers/Chat/Chat.tsx:72]: TODO here

**Review by AndyTWF — COMMENTED** [review]
We need to add tests and README here (appreciate we're still in draft, so just making a note)

File: test/react/hooks/use-messages.integration.test.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      }
+    }, 100);
+    setTimeout(() => {
+      clearInterval(interval);
+      reject(new Error('Timed out waiting for messages'));
+    }, 5000);
+  });
+}
+
+describe('useMessages', () => {
```
Comment (AndyTWF) [inline test/react/hooks/use-messages.integration.test.tsx:27]: Can we have a test here for `getPreviousMessages`

File: src/react/README.md [hunk 3]
```diff
@@ -127,6 +127,73 @@ const MyComponent = () => {
...
+    listener: (message) => {
+      console.log('Received message: ', message);
+    },
+    // optional parameter to fetch previous messages up until listener was subscribed
+    previousMessagesParams: { limit: 10 },
+  });
+
+  useEffect(() => {
+    // fetch last 10 messages up until listener was subscribed
+    if (previousMessages) {
```
[the remark this answers is not in the document]
Comment (splindsay-92) [reply src/react/README.md:188]: Good idea, I will add this in, and also add this to the demo.

File: src/react/hooks/use-messages.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   */
+  readonly messages: Messages;
+
+  /**
+   * If a listener is provided in {@link UseMessagesParams},
+   * this promise will resolve to a paginated result of messages fetched
+   * based on the {@link UseMessagesParams.previousMessagesParams}.
+   * The results will be in the newest-to-oldest order.
+   */
+  readonly previousMessages?: Promise<PaginatedResult<Message>>;
```
[the remark this answers is not in the document]
Comment (splindsay-92) [reply src/react/hooks/use-messages.ts]: I have updated this to return a function.

File: demo/src/containers/Chat/Chat.tsx [hunk 5]
```diff
@@ -1,36 +1,58 @@
...
+  const isConnected: boolean = currentStatus === ConnectionLifecycle.Connected;
+
+  const { send: sendMessage, getPreviousMessages } = useMessages({
+    listener: (message) => {
+      setMessages((prevMessage) => [...prevMessage, message.message]);
+    },
+    onDiscontinuity: (discontinuity) => {
+      console.log('Discontinuity', discontinuity);
+      // reset the messages when a discontinuity is detected,
+      // this will trigger a re-fetch of the messages
```
Comment (vladvelici) [inline demo/src/containers/Chat/Chat.tsx:25]: ambiguity: discontinuity triggers the re-fetch or `setMessages([])`?

File: demo/src/containers/Chat/Chat.tsx [hunk 6]
```diff
@@ -1,36 +1,61 @@
...
-      sendMessage(text);
-    },
-    [sendMessage],
-  );
+  useEffect(() => {
+    // try and fetch the messages up to attachment of the messages listener
+    if (getPreviousMessages && loading) {
+      getPreviousMessages({ limit: 50 })
+        .then((result) => {
+          setMessages(result.items);
```
Comment (vladvelici) [inline demo/src/containers/Chat/Chat.tsx:50]: reverse the array? seems like it shows the wrong way around (now it's top-to-bottom newest-to-oldest)

File: src/react/hooks/use-messages.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      // set the getPreviousMessages method if a listener is provided
+      setGetPreviousMessages(() => {
+        logger.debug('useMessages(); setting getPreviousMessages state', { roomId: room.roomId });
+        return (params: Omit<QueryOptions, 'direction'>) => {
+          return sub.getPreviousMessages(params);
+        };
+      });
+
+      return () => {
+        logger.debug('useMessages(); removing listener', { roomId: room.roomId });
```
Comment (vladvelici) [inline src/react/hooks/use-messages.ts:101]: nitpick: should this and the log on line 103 be a single log? or trace one of them or both? seems like a lot of logging

Comment (splindsay-92) [reply src/react/hooks/use-messages.ts:101]: Good point, I'll combined this to a single log

File: demo/src/main.tsx [hunk 8]
```diff
@@ -40,7 +40,7 @@ const realtimeClient = new Ably.Realtime({
   clientId,
 });
 
-const chatClient = new ChatClient(realtimeClient, { logLevel: LogLevel.Debug });
+const chatClient = new ChatClient(realtimeClient, { logLevel: LogLevel.Debug, logHandler: console.log });
```
Comment (AndyTWF) [inline demo/src/main.tsx:43]: FWIW the default is to use an inbuilt logger based on console.log, which also includes properly using `console.error` in error situations.

Should we let it use the default here?

File: src/react/contexts/chat-client-context.tsx [hunk 9]
```diff
@@ -2,8 +2,8 @@ import { ChatClient } from '@ably/chat';
 import React from 'react';
 
 /**
- * Context key to handle global context for Ably Chat Client.
- * Only a single instance of Ably Chat context should exist to avoid issues resulted from multiple identical contexts,
+ * THe Context key to handle global context for Ably Chat Client.
```
Comment (AndyTWF) [inline src/react/contexts/chat-client-context.tsx:5]:
```suggestion
 * The Context key to handle global context for Ably Chat Client.
```

File: src/react/hooks/use-messages.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const logger = useLogger();
+  logger.trace('useMessages();', { params, roomId: room.roomId });
+
+  // we are storing the params in a ref so that we don't end up with an infinite loop should the user pass
+  // in an unstable reference
+  const paramsRef = useRef(params);
```
Comment (AndyTWF) [inline src/react/hooks/use-messages.ts:76]: Depending on who gets to merge first, will update this to use the new function

**Review by AndyTWF — COMMENTED** [review]
Final nitpicks :)

**Review by AndyTWF — APPROVED** [review]

**Review by vladvelici — APPROVED** [review]

**Pull request merged**